### PR TITLE
neovim: generate remote plugin manifest

### PR DESCRIPTION
**Pull request by timokau**
###### Motivation for this change

This makes sure the user doesn't have to call `UpdateRemotePlugins` manually for plugins installed through nix. A minor patch to neovim is necessary, but it should be harmless. See neovim/neovim#9413 for a discussion about the patch.

cc @Mic92 @rvolosatovs 

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by timokau** [comment]
I tested this with the `deoplete-nvim` plugin. The user can still additionally use `UpdateRemotePlugins` as before.

File: pkgs/applications/editors/neovim/wrapper.nix [hunk 1]
```diff
@@ -83,18 +82,29 @@ let
...
+      '' + optionalString (configure != {}) ''
+        echo "Generating remote plugin manifest"
+        export NVIM_RPLUGIN_MANIFEST=$out/rplugin.vim
+        if ! $out/bin/nvim -N -u ${vimUtils.vimrcFile (configure // { customRC = ""; })} -i NONE -n -E -s +UpdateRemotePlugins +quit!; then
+          echo "Generating rplugin.vim failed!"
+          exit 1
+        fi
+        unset NVIM_RPLUGIN_MANIFEST
+
+        # this relys on a patched neovim, see
```
Comment (rvolosatovs) [inline pkgs/applications/editors/neovim/wrapper.nix]: `relays`

**Review by rvolosatovs — APPROVED** [review]
Pure awesomeness - great work!
I haven't tested it yet, but I can't come up with a reason why it would not work

File: pkgs/applications/editors/neovim/wrapper.nix (hunk 1 above)
Comment (timokau) [reply pkgs/applications/editors/neovim/wrapper.nix]: I meant to say relies, always get confused with the ie. Fixed :)

**Comment by timokau** [comment]
Given the recent fallout on my last refactor (https://github.com/NixOS/nixpkgs/issues/53112, https://github.com/NixOS/nixpkgs/pull/53113) I've added some better error handling and a testcase (we should have more testcases). The error handling is important since we're starting neovim with a vimrc. That vimrc *should* not contain any errors since it was automatically generated, but  in case it does this will make it easier to debug.

I've also added the error handling (e.g. not silently ignoring errors) to the helptag generation, although that is less likely to fail.

I have
- run `nix-review` to make sure all plugins and all testcases build
- verified deoplete works with and without `~/.local/share/nvim/rplugin.vim` present

**Comment by timokau** [comment]
@GrahamcOfBorg build vimUtils.test_nvim_with_remote_plugin

**Comment by timokau** [comment]
The high rebuild count is because `buildVimPlugin` was touched. However the vim plugins are all extremely fast to "build".

**Comment by Mic92** [comment]
Do you plan to open a pull request for that?

**Comment by timokau** [comment]
Oops, totally forgot this PR was still open. Thanks for the reminder @teto. I'll rebase this.

@Mic92 will who open a PR for what?

**Comment by timokau** [comment]
@GrahamcOfBorg build vimUtils.test_nvim_with_remote_plugin

**Comment by timokau** [comment]
@Mic92 any objections to merge?

**Comment by rvolosatovs** [comment]
FTR, I've been running this for a while now and it seems to work fine.

**Review by Mic92 — APPROVED** [review]

**Pull request merged**